cpu/atmega_common/periph/adc.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2016 Laurent Navet <laurent.navet@gmail.com>
+ *               2017 HAW Hamburg, Dimitri Nahm
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level ADC driver implementation for ATmega family
+ *
+ * @author      Laurent Navet <laurent.navet@gmail.com>
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "assert.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+
+#define ADC_MAX_CLK         (200000U)
+
+static mutex_t lock = MUTEX_INIT;
+
+static inline void prep(void)
+{
+    mutex_lock(&lock);
+    /* Enable ADC */
+    ADCSRA |= (1 << ADEN);
+}
+
+static inline void done(void)
+{
+    /* Disable ADC */
+    ADCSRA &= ~(1 << ADEN);
+    mutex_unlock(&lock);
+}
+
+int adc_init(adc_t line)
+{
+    uint32_t clk_div = 1;
+
+    /* check if the line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    prep();
+
+    /* Disable corresponding Digital input */
+    #if defined (CPU_ATMEGA328P) || defined (CPU_ATMEGA1281)
+        DIDR0 |= (1 << line);
+    #elif defined (CPU_ATMEGA2560)
+        if (line < 8)
+            DIDR0 |= (1 << line);
+        else
+            DIDR2 |= (1 << (line - 8));
+    #endif
+
+    /* Set ADC-pin as input */
+    #if defined (CPU_ATMEGA328P)
+        DDRC &= ~(1 << line);
+        PORTC &= ~(1 << line);
+    #elif defined (CPU_ATMEGA2560)
+        if (line < 8) {
+            DDRF &= ~(1 << line);
+            PORTF &= ~(1 << line);
+        }
+        else {
+            DDRK &= ~(1 << (line-8));
+            PORTK &= ~(1 << (line-8));
+        }
+    #elif defined (CPU_ATMEGA1281)
+        DDRF &= ~(1 << line);
+        PORTF &= ~(1 << line);
+    #endif
+
+    /* set clock prescaler to get the maximal possible ADC clock value */
```

Comment:
Idea: since prescalers just depends in the coreclock, can be set by the pre-processor
```
/* sampling freq must be between 200kHz and 50kHz */
#if CLOCK_CORECLOCK >  12800000
#define PRESCALER 0x07  /* 128 */
#elif CLOCK_CORECLOCK > 6400000
#define PRESCALER 0x06  /* 64 */
#elif CLOCK_CORECLOCK > 3200000
#define PRESCALER 0x05  /* 32 */
#elif CLOCK_CORECLOCK > 1600000
#define PRESCALER 0x04  /* 16 */
#elif CLOCK_CORECLOCK >  800000
#define PRESCALER 0x03  /* 8 */
#elif CLOCK_CORECLOCK >  100000
#define PRESCALER 0x01  /* 2 */
#else
#error CLOCK must be at least 100kHz to use ADC
#endif
```

Reply:
in #7502 its the other way around, in general I'd say the compiler should optimise that out anyway. But to be sure, you could make a size comparison. 

However, I would leave as is for now.